items/src/main/scala/weco/api/items/services/ItemUpdateService.scala

```diff
@@ -23,40 +26,57 @@ class ItemUpdateService(
   private def isSierraId(sourceIdentifier: SourceIdentifier) =
     sourceIdentifier.identifierType == IdentifierType.SierraSystemNumber
 
-  private def updateAccessCondition(
+  private def updateLocations(
     item: Item[IdState.Minted],
     accessCondition: AccessCondition
-  ) =
-    item.locations.map {
-      case physicalLocation: PhysicalLocation =>
-        physicalLocation.copy(
-          accessConditions = List(accessCondition)
-        )
-      case location => location
-    }
+  ) = item.locations.map {
+    case physicalLocation: PhysicalLocation =>
+      physicalLocation.copy(
+        accessConditions = List(accessCondition)
+      )
+    case location => location
+  }
 
-  private def refreshItem(srcId: SourceIdentifier, item: Item[IdState.Minted]) =
-    sierraService
-      .getAccessCondition(srcId)
-      .map {
-        case Right(Some(accessCondition)) =>
-          item.copy(
-            locations = updateAccessCondition(item, accessCondition)
-          )
-        case Right(_) => item
-        case Left(err) =>
-          error(msg = f"Couldn't refresh item: ${item.id} got error $err")
-          item
+  private def updateAccessConditions(
+    sierraItemSourceIdentifiers: Map[SierraItemNumber, Item[IdState.Minted]],
+    accessConditions: Map[SierraItemNumber, Option[AccessCondition]]
+  ): immutable.Iterable[Item[IdState.Minted]] = {
+    accessConditions.flatMap {
+      case (itemNumber, Some(accessCondition)) => {
+        sierraItemSourceIdentifiers
+          .get(itemNumber)
+          .map(item => {
+            item.copy(
+              locations = updateLocations(item, accessCondition)
+            )
+          })
       }
+    }
+  }
 
   def updateItems(
     work: Work.Visible[WorkState.Indexed]
-  ): Future[List[Item[IdState.Minted]]] = Future.sequence {
-    work.data.items.map {
-      case item @ Item(IdState.Identified(_, srcId, _), _, _, _)
-          if isSierraId(srcId) =>
-        refreshItem(srcId, item)
-      case item => Future(item)
+  ): Future[Iterable[Item[IdState.Minted]]] = {
+    val (sierraItems, otherItems) = work.data.items.partition {
+      case Item(IdState.Identified(_, srcId, _), _, _, _) =>
+        isSierraId(srcId)
     }
+
+    val sierraItemSourceIdentifiers = sierraItems.map {
+      case item @ Item(IdState.Identified(_, srcId, _), _, _, _) =>
+        SierraItemIdentifier.fromSourceIdentifier(srcId) -> item
+    } toMap
+
+    val updatedSierraItems = sierraService
+      .getAccessConditions(sierraItemSourceIdentifiers.keys.toSeq)
+      .map {
+        case Right(accessConditions) =>
+          updateAccessConditions(sierraItemSourceIdentifiers, accessConditions)
+        case Left(err) =>
+          error(msg = f"Couldn't refresh items on: ${work.id} got error $err")
+          sierraItems
+      }
+
+    updatedSierraItems.map(_ ++ otherItems)
```

Comment:
You linked to this function as particularly knotty. Perhaps it would help to think of Sierra as 1 of multiple sources we might implement requesting for, rather than a special case?

I was about to suggest some sort of match statement that goes like this:

```scala
items.map {
  case sierraItem => updateSierraItem(sierraItem)
  case calmItem => updateCalmItem(calmItem)
  case item => Future.successful(item)
}
```

but of course, the whole point of this work is to update items in bulk!

How about something like this:

```scala
case class Item(id: String, scheme: String)

val items = Seq(
  Item(id = "b123", scheme = "Sierra"),
  Item(id = "b124", scheme = "Sierra"),
  Item(id = "A/B/C", scheme = "Calm"),
  Item(id = "b125", scheme = "Sierra"),
  Item(id = "V0000001", scheme = "Miro")
)

val itemsByScheme = items.groupBy { _.scheme }
// HashMap(
//   Miro -> List(Item(V0000001,Miro)),
//   Calm -> List(Item(A/B/C,Calm)),
//   Sierra -> List(Item(b123,Sierra), Item(b124,Sierra), Item(b125,Sierra))
// )

itemsByScheme.map {
  case (Sierra, sierraItems) => updateSierraItems(sierraItems)
  case (Calm, calmItems) => updateCalmItems(calmItems)
  case (_, items) => items
}
```

I might also throw in a `zipWithIndex` somewhere to preserve the original item ordering, which currently is being lost.